webapp/src/ts/services/format-data-record.service.ts

```diff
@@ -0,0 +1,716 @@
+  *          },
+  *          ...
+  *        ],
+  *        sent_by: 'admin',
+  *        from: '+998',
+  *        state: 'sent',
+  *        message: 'good morning'
+  *    }
+  *  ]
+  */
+  private formatOutgoingMessages(doc) {
+    const outgoing_messages = [];
+    const outgoing_messages_recipients = [];
+    doc.tasks.forEach((task) => {
+      task.messages.forEach((msg) => {
+        const recipient = {
+          to: msg.to,
+          facility: msg.facility,
+          timestamp: task.timestamp,
+          uuid: msg.uuid,
+        };
+        let done = false;
+        // append recipient to existing
+        outgoing_messages.forEach((m) => {
+          if (
+            msg.message === m.message &&
+            msg.sent_by === m.sent_by &&
+            msg.from === m.from &&
+            task.state === m.state
+          ) {
+            m.recipients.push(recipient);
+            outgoing_messages_recipients.push(recipient);
+            done = true;
+          }
+        });
+        // create new entry
+        if (!done) {
+          outgoing_messages.push({
+            recipients: [recipient],
+            sent_by: msg.sent_by,
+            from: msg.from,
+            state: task.state,
+            message: msg.message,
+          });
+          outgoing_messages_recipients.push(recipient);
+        }
+      });
+    });
+    doc.outgoing_messages = outgoing_messages;
+    doc.outgoing_messages_recipients = outgoing_messages_recipients;
+  }
+
+  /*
+    * Take data record document and return nice formated JSON object.
+  */
+  private makeDataRecordReadable(doc, settings, language, context) {
+    const formatted = _.clone(doc);
+
+    if (formatted.content_type === 'xml') {
+      this.formatXmlFields(formatted);
+    } else {
+      this.formatJsonFields(formatted, settings, language);
+    }
+
+    if (formatted.scheduled_tasks) {
+      this.formatScheduledTasks(formatted, settings, language, context);
+    }
+
+    if (formatted.kujua_message) {
+      this.formatOutgoingMessages(formatted);
+    }
+
+    return formatted;
+  }
+
+  format(doc) {
+    const promises = [this.settingsService.get(), this.languageService.get()];
+    const patientId = doc.patient_id || (doc.fields && doc.fields.patient_id);
+    if (doc.scheduled_tasks && patientId) {
+      promises.push(this.getPatient(patientId));
+      promises.push(this.getRegistrations(patientId));
+    }
+    return Promise.all(promises).then((results) => {
+      const settings = results[0];
+      const language = results[1];
+      const context:any = {};
+      if (results.length === 4) {
+        context.patient = results[2];
+        context.registrations = (<[]>results[3]).filter((registration) => {
+          return registrationUtils.isValidRegistration(
+            registration,
+            settings
+          );
+        });
+      }
+      return this.makeDataRecordReadable(doc, settings, language, context);
+    });
+  }
+}
```

Comment:
[Cries in Spanish] 
I thought I'd never stop scrolling this file haha

Reply:
Yea. I know. Did you read every line? :))))